### PR TITLE
Explore 'module.run' state module output in depth to catch "result" properly

### PR DESCRIPTION
**Pull request by meaksh**
### What does this PR do?
This PR adds a mechanism into the `module.run` state module to explore in depth the output dictionary returned by the called module in order to find out the `result` attribute (which in some cases can be wrapped by the output of the module called).

Currently, when the `state.apply` module is called within a `module.run` state, the response produced by the `module.run` does not properly set the `result` attribute. Therefore the `require` statements are not working on these cases.

Given the following SLS file:
```yaml
my_test_state_1:
  module.run:
    - name: state.apply
    - mods: my_other_state

my_test_state_2:
  module.run:
    - name: state.apply
    - mods: my_other_state_2
    - require:
      - module: my_test_state_1
```

### Previous Behavior
```
minsles12sp3-demo-1.tf.local:                                                                                                                                               
----------                                                                                                                                                                  
          ID: my_test_state_1
    Function: module.run
        Name: state.apply
      Result: True
     Comment: Module function state.apply executed
     Started: 11:12:01.866882
    Duration: 130.021 ms
     Changes:   
              ----------
              ret:
                  ----------
                  test_|-execute_command_|-test_|-fail_with_changes:
                      ----------
                      __id__:
                          execute_command
                      __run_num__:
                          0
                      changes:
                          ----------
                          testing:
                              ----------
                              new:
                                  Something pretended to change
                              old:
                                  Unchanged
                      comment:
                          Failure!
                      duration:
                          1.832
                      name:
                          test
                      result:
                          False
                      start_time:
                          11:12:01.992528
----------
          ID: my_test_state_2
    Function: module.run
        Name: state.apply
      Result: True
     Comment: Module function state.apply executed
     Started: 11:12:01.997184
    Duration: 101.492 ms
     Changes:   
              ----------
              ret:
                  ----------
                  test_|-execute_command_|-test_|-nop:
                      ----------
                      __id__:
                          execute_command
                      __run_num__:
                          0
                      changes:
                          ----------
                      comment:
                          Success!
                      duration:
                          0.238
                      name:
                          test
                      result:
                          True
                      start_time:
                          11:12:02.097643

Summary for minsles12sp3-demo-1.tf.local
------------
Succeeded: 2 (changed=2)
Failed:    0
------------
Total states run:     2
Total run time: 231.513 ms
```

As you can see, the result for `my_test_state_1` is set to `True` while the `ret` output returned by the `state.apply` module called contains `result = False`. Therefore the two states are actually executed while the expected behavior would be that the second state won't be executed due the `require` statement.

### New Behavior
```
minsles12sp3-demo-1.tf.local:
----------         
          ID: my_test_state_1                                     
    Function: module.run
        Name: state.apply
      Result: False                     
     Comment: Module function state.apply executed
     Started: 11:15:55.548039
    Duration: 86.597 ms
     Changes:   
              ----------
              ret:        
                  ----------
                  test_|-execute_command_|-test_|-fail_with_changes:
                      ----------
                      __id__:
                          execute_command
                      __run_num__:
                          0
                      changes:
                          ----------
                          testing:
                              ----------
                              new:
                                  Something pretended to change
                              old:
                                  Unchanged
                      comment:
                          Failure!
                      duration:
                          1.865
                      name:
                          test
                      result:
                          False
                      start_time:
                          11:15:55.629306
----------
          ID: my_test_state_2
    Function: module.run
        Name: state.apply
      Result: False
     Comment: One or more requisite failed: testpr.my_test_state_1
     Changes:   

Summary for minsles12sp3-demo-1.tf.local
------------
Succeeded: 0 (changed=1)
Failed:    2
------------
Total states run:     2
Total run time:  86.597 ms
ERROR: Minions returned with non-zero exit code

```

### Tests written?

Yes

### Commits signed with GPG?

Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by meaksh** [comment]
This still needs some fixes. Setting as [WIP].

File: salt/states/module.py [hunk 1]
```diff
@@ -530,7 +530,21 @@ def _get_result(func_ret, changes):
...
 
+
+def _get_result_in_depth(node):
+    for key, val in six.iteritems(node):
+        if key == 'result' and val is False:
+            return False
+        elif isinstance(val, dict):
+            return _get_result_in_depth(val)
+    return True
+
```
Comment (isbm) [inline salt/states/module.py]: Please make sure this applies to `def run` as well and please write Unit Tests  for this. Thanks!

**Review by isbm — CHANGES_REQUESTED** [review]

**Comment by meaksh** [comment]
This new function is called inside the `_get_result()` which is called from both `module.run` and `module._run` methods to gather the results from the response of the called module.

I've fixed the recursive method + added unit test. Removing [WIP] tag now. 

@isbm @rallytime, this is now ready for review. Thanks. :+1:

**Review by isbm — APPROVED** [review]

**Review by rallytime — APPROVED** [review]

**Comment by rallytime** [comment]
re-run py3

**Review by garethgreenaway — APPROVED** [review]

**Pull request merged**